### DISCUSSION TITLE
Prevent sliding down slopes when using physics & gravity

### DISCUSSION THREAD

**Discussion by mattjennings** [excaliburjs/Excalibur#2459] [Q&A]
How might I go about allowing an object to move up & down a slope without sliding down due to gravity? See this example:

https://codesandbox.io/s/excalibur-slopes-qgi0cz?file=/src/index.js

Try walking onto the slope with the arrow keys. Notice how the box slides down when you stop moving, and is also moving at different speeds when going up/down due to gravity.

I have tried increasing the friction on the body to a very high number but that didn't seem to affect anything. I also tried disabling gravity for the body when in contact with the ground, and while this works when walking up, walking down of course requires gravity being enabled again otherwise the object floats off of it.

At this point, I feel like disabling gravity when on the ground & finding a way to enable gravity when walking down the slope is my best option, but I'm curious if there are any other solutions.

**Comment by ewal** [comment]
The first idea that came to my mind was also to disable gravity when grounded and if `vel.x` equals to zero, but there might of course be a neater way of handling this.

**Reply by mattjennings** [reply]
Oh that's a good idea! That would solve sliding down while standing still at least. There would still be different movement speeds when walking up / down, but that is potentially a separate issue (although still related to gravity being the culprit).

**Reply by eonarheim** [reply]
I think that's a solid idea, I'm trying that toggle gravity approach out locally. It's possible to set the Player actor acceleration specifically if you don't want to toggle the global gravity

**Reply by ewal** [reply]
If the angle is known or could be read/emitted from the collision object perhaps it might be possible to apply the angle to the x factor in the gravity vector, if gravity is handled in the player actor that is. That would solve the sliding issue I'd guess.

**Reply by eonarheim** [reply]
This seems to partially work, there is some wiggle still for some reason (using [option 2](https://github.com/excaliburjs/Excalibur/discussions/2406#discussioncomment-3117898) from the platforming discussion might be more robust). 

When raycasting (#2401) is merged that might be another way to more elegantly handle slopes.

```typescript
class Player extends ex.Actor {
...
  onInitialize() {
    this.on("precollision", (event) => {
      this.playerSideCollisions.push(event.side);
    });
  }
  isPlayerOnGround() {
    return this.playerSideCollisions.includes(ex.Side.Bottom);
  }
  onPostUpdate(engine) {
    if (this.isPlayerOnGround()) {
      // console.log("onground");
      this.acc = ex.vec(0, 0);
    } else {
      // console.log("not ground");
      this.acc = ex.vec(0, 10000);
    }
    this.vel = ex.vec(0, 0);

    if (engine.input.keyboard.isHeld(ex.Input.Keys.ArrowLeft)) {
      this.vel.x = -200;
    }

    if (engine.input.keyboard.isHeld(ex.Input.Keys.ArrowRight)) {
      this.vel.x = 200;
    }
    this.playerSideCollisions.length = 0; // clear last frame's side collisions
  }
}

```

**Reply by mattjennings** [reply]
@eonarheim I haven't had a chance to try this yet but thank you for this! A combination of detecting slopes through raycasting and disabling gravity very well might do the trick. I'll probably circle back to this once that PR is merged in and see what I can come up with.
